Create RR.ProviderData (close #119)

mholt opened the pull request:
This is my proposal for being able to attach provider-specific (meta?)data to an RR. See #119.

However, since provider implementations are supposed to return the concrete struct types, not RRs, we'll likely need to add this field to all the specific structs (Address, TXT, CNAME, etc...).

I also added JSON tags to the RR struct, in preparation for 1.0, since JSON-serialization seems like it could be a common pattern. (Caddy/CertMagic do not do this, but I could see having separate JSON identifiers being handy.)

Finally, I am torn between `Meta`, `Metadata`, and `ProviderData` for the field name. Any thoughts?

@gucci-on-fleek What do you think?

mholt commented:
(Super skeptical of an AI review but I'm curious, so...)

pcfreak30 commented:
seems simple enough, and btw AI is good enough to find issues (I use it near daily), but very far away from replacing us :P.

@mholt i would add `omitempty` and i think this is good.

mholt commented:
> i would add omitempty and i think this is good.

Thanks; I think only `ProviderData` (as currently named) would ever be empty... I think? Maybe TTL, depending on when/where provider defaults are applied.

gucci-on-fleek commented:
> I also added JSON tags to the RR struct, in preparation for 1.0, since JSON-serialization seems like it could be a common pattern.

:+1:

> Finally, I am torn between `Meta`, `Metadata`, and `ProviderData` for the field name. Any thoughts?

`ProviderData` seems like the least-worst name that I can think of. It's not an amazing name, but certainly better than anything else that I can think of.

> However, since provider implementations are supposed to return the concrete struct types, not RRs, we'll likely need to add this field to all the specific structs (Address, TXT, CNAME, etc...).

Hmm, yeah, you're probably right here. So then calling `.Parse()` or `.RR()` should probably just copy over the contents of `ProviderData` unmodified? I guess the other options would be to provide a hook for this or unconditionally set `ProviderData` to `nil`, but neither of those seem great to me.

Or, maybe it would be better to _only_ have `ProviderData` on the specific structs, and not on `RR`? My thinking being that you can change the `Type` of an `RR`, and providers might handle that very poorly. I don't think that this should be very much of a problem though, just something to think about.

> @gucci-on-fleek What do you think?

The main problem that I can see is that we'll either need to document that `RR`s aren't comparable

https://github.com/libdns/libdns/blob/10a2b0b2c222c51900827339aaf1e22f5cfd2528/record.go#L13-L15

which might be annoying since that means that there's no “easy” way to compare records, or document that `ProviderData` must be primitively-comparable, which would also be annoying since then you couldn't use it to hold a `map`.

The other problem that I can think of is that weird things might happen if you try and copy `Record`s between different providers, so we should probably document either that `libdns` implementations must ignore invalid/unknown `ProviderData`s, or that users must set `ProviderData` to `nil` when copying between zones/providers.

But overall, I'm very much in favour of this.

mholt commented:
@gucci-on-fleek Okay, I reworked it so that ProviderData is in each individual struct type, not the RR struct. I also added some package-level godoc surrounding it.

Since packages are supposed to return the individual struct types anyway, they can just attach their custom data in this field, and callers' type assertions will still work. Obviously, the provider-specific data is lost with `RR()` (and thus, `Parse()` does not populate it), but that's OK since the `RR` type is meant to be more generic/portable, and comparable.

gucci-on-fleek commented on libdns.go:
```diff
@@ -19,6 +19,15 @@
+// Record values should not be primitvely compared (==) unless it is [RR],
+// because some struct types contain maps, for which equality is not defined;
+// additionally, some packages may attach custom data to each RR struct-type's
+// `ProviderData` field, which values might not be comparable either. The
```
```suggestion
// `ProviderData` field, whose values might not be comparable either. The
```

gucci-on-fleek reviewed (approved):
LGTM, hopefully this helps make it easier to port implementations to v1.0.

mholt commented:
Awesome, thanks!
